### PR TITLE
Support lambda expressions

### PR DESCRIPTION
**Pull request by ArsenHD**
# Description

Substitute this text with a concise description of the proposed change. Emphasize, why particular solution was chosen.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- Minor bug fix (non-breaking small changes)
- Bug fix (non-breaking change which fixes an issue)
- Refactoring (typos and non-functional changes) 
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

## Automated Testing

Specify tests that help to verify the change automatically.  

_Example:_ org.utbot.examples.algorithms.BinarySearchTest

## Manual Scenario 

Please, provide several scenarios that you went through to verify that the change worked as expected.  

# Checklist (remove irrelevant options):

_This is the author self-check list_

- [x] The change followed the style guidelines of the UTBot project
- [x] Self-review of the code is passed
- [x] The change contains enough commentaries, particularly in hard-to-understand areas
- [x] New documentation is provided or existed one is altered
- [x] No new warnings
- [x] New tests have been added
- [ ] All tests pass locally with my changes

### CODE REVIEW

File: utbot-framework/src/main/kotlin/org/utbot/engine/util/lambda/LambdaConstructionUtils.kt [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    // Obtain the single abstract method of a functional interface whose instance we are building.
+    // For example, for `java.util.function.Predicate` it will be method `test`.
+    val singleAbstractMethod = getSingleAbstractMethod(samType)
+    val invokedName = singleAbstractMethod.name
+    // Method type of single abstract method of the target functional interface.
+    val samMethodType = MethodType.methodType(singleAbstractMethod.returnType, singleAbstractMethod.parameterTypes)
+
+    val lambdaMethod = getLambdaMethod(declaringClass, lambdaName)
+    lambdaMethod.isAccessible = true
+    val lambdaMethodType = MethodType.methodType(lambdaMethod.returnType, lambdaMethod.parameterTypes)
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/engine/util/lambda/LambdaConstructionUtils.kt:144]: Create a class for this code with restructuring, please

File: utbot-framework/src/main/kotlin/org/utbot/engine/Traverser.kt [hunk 2]
```diff
@@ -1831,7 +1830,7 @@ class Traverser(
             return fromMemory
         }
         val addr = fromMemory?.addr ?: findNewAddr()
-        val created = createObject(addr, classType, useConcreteType = false, mockInfoGenerator)
+        val created = createObject(addr, classType, useConcreteType = true, mockInfoGenerator)
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/engine/Traverser.kt:1833]: By mistake?

File: utbot-framework/src/main/kotlin/org/utbot/engine/TypeResolver.kt [hunk 3]
```diff
@@ -208,8 +211,15 @@ class TypeResolver(private val typeRegistry: TypeRegistry, private val hierarchy
...
-                    sootClass.isArtificialEntity -> if (keepArtificialEntities) concreteTypes += it else Unit
+//                    sootClass.isAnonymous || sootClass.isUtMock -> unwantedTypes += it
+                    sootClass.isArtificialEntity -> {
+//                        if (sootClass.isLambda) {
+//                            unwantedTypes += it
+//                        } else
+                        if (keepArtificialEntities) {
+                            concreteTypes += it
+                        }
+                    }
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/engine/TypeResolver.kt:222]: Fix this method, please

File: utbot-framework/src/main/kotlin/org/utbot/engine/TypeResolver.kt [hunk 4]
```diff
@@ -113,7 +113,7 @@ class TypeResolver(private val typeRegistry: TypeRegistry, private val hierarchy
                 if (numDimensions == 0) baseType else baseType.makeArrayType(numDimensions)
             }
 
-        return TypeStorage(type, concretePossibleTypes).filterInappropriateClassesForCodeGeneration()
+        return TypeStorage(type, concretePossibleTypes)//.filterInappropriateClassesForCodeGeneration()
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/engine/TypeResolver.kt:116]: Uncomment?

File: utbot-framework/src/main/kotlin/org/utbot/engine/TypeResolver.kt [hunk 5]
```diff
@@ -183,7 +183,7 @@ class TypeResolver(private val typeRegistry: TypeRegistry, private val hierarchy
             else -> error("Unexpected type $type")
         }
 
-        return TypeStorage(type, possibleTypes).filterInappropriateClassesForCodeGeneration()
+        return TypeStorage(type, possibleTypes)//.filterInappropriateClassesForCodeGeneration()
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/engine/TypeResolver.kt:186]: Uncomment?

File: utbot-framework/src/main/kotlin/org/utbot/engine/UtBotSymbolicEngine.kt [hunk 6]
```diff
@@ -473,14 +473,15 @@ class UtBotSymbolicEngine(
...
-            }
+                // do not remove concrete results of anonymous classes, because now we are able to generate code for them
+//                workaround(REMOVE_ANONYMOUS_CLASSES) {
+//                    concreteExecutionResult.result.onSuccess {
+//                        if (it.classId.isAnonymous) {
+//                            logger.debug("Anonymous class found as a concrete result, symbolic one will be returned")
+//                            return@flow
+//                        }
+//                    }
+//                }
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/engine/UtBotSymbolicEngine.kt:484]: Remove?

File: utbot-framework/src/main/kotlin/org/utbot/engine/UtBotSymbolicEngine.kt [hunk 7]
```diff
@@ -591,15 +592,15 @@ class UtBotSymbolicEngine(
...
-                }
+//                workaround(REMOVE_ANONYMOUS_CLASSES) {
+//                    concreteExecutionResult.result.onSuccess {
+//                        if (it.classId.isAnonymous) {
+//                            logger.debug("Anonymous class found as a concrete result, symbolic one will be returned")
+//                            emit(symbolicUtExecution)
+//                            return
+//                        }
+//                    }
+//                }
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/engine/UtBotSymbolicEngine.kt:603]: Remove?

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/visitor/UtilMethods.kt [hunk 8]
```diff
@@ -55,11 +62,32 @@ internal fun UtilMethodProvider.utilMethodTextById(
...
+            getSingleAbstractMethodMethodId -> getSingleAbstractMethod(codegenLanguage)
             else -> error("Unknown util method for class $this: $id")
         }
     }
 }
 
+internal fun UtilMethodProvider.auxiliaryClassTextById(
+    id: ClassId,
+    codegenLanguage: CodegenLanguage
+): Result<String> = runCatching {
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/visitor/UtilMethods.kt:82]: Remove `Result` here and throw an error here, please

Also, please create a task for isolation of rendering test method and link it here

**Review by Damtev — CHANGES_REQUESTED** [review]

File: utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/Api.kt [hunk 9]
```diff
@@ -529,6 +530,35 @@ data class UtAssembleModel(
...
+ * Lambdas in Java represent the implementation of a single abstract method (SAM) of a functional interface.
+ * They can be used to create an instance of said functional interface, but **they are not classes**.
+ * In Java lambdas are compiled into synthetic methods of a class they are declared in.
+ * Depending on the captured variables, this method will be either static or non-static.
+ *
+ * Since lambdas are not classes we cannot use a class loader to get info about them as we can do for other models.
+ * Hence, the necessity for this specific lambda model that will be processed differently: instead of working
+ * with a class we will be working with the synthetic method that represents our lambda.
+ */
+// TODO: what about support for Kotlin lambdas (they are not exactly the same as Java's due to functional types)
```
[the remark this answers is not in the document]
Comment (ArsenHD) [reply utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/Api.kt:545]: https://github.com/UnitTestBot/UTBotJava/issues/852

File: utbot-framework/src/main/kotlin/org/utbot/engine/util/lambda/LambdaConstructionUtils.kt [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+ * @return [MethodType] that represents the value of argument `instantiatedMethodType`
+ * of method [LambdaMetafactory.metafactory].
+ */
+private fun getInstantiatedMethodType(
+    lambdaMethod: Method,
+    capturedArgumentTypes: Array<Class<*>>
+): MethodType {
+    // Types of arguments of synthetic method (representing lambda) without the types of captured values.
+    val instantiatedMethodParamTypes = lambdaMethod.parameterTypes
+        .drop(capturedArgumentTypes.size)
```
[the remark this answers is not in the document]
Comment (ArsenHD) [reply utbot-framework/src/main/kotlin/org/utbot/engine/util/lambda/LambdaConstructionUtils.kt:48]: Turns out `lambdaMethod.parameterTypes` is an array, so there is no way to obtain a view on some portion of it. That's why we will have to leave it like that.

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/visitor/UtilMethods.kt (hunk 8 above)
Comment (ArsenHD) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/visitor/UtilMethods.kt:82]: Done. Here's the [taks](https://github.com/UnitTestBot/UTBotJava/issues/853)

**Review by Damtev — APPROVED** [review]
LGTM

**Pull request merged**